### PR TITLE
Reject the creation of interface with the new statement

### LINKED ISSUE #2733: new expression with library and interface "works"
```
interface I {
}

library L {
}

contract C {
  function f() {
    new I();
  }

  function g() {
    new L();
  }
}
```

Not sure about creating libraries, but creating interfaces must be disallowed.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2733.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1543,6 +1543,8 @@ void TypeChecker::endVisit(NewExpression const& _newExpression)
 
 		if (!contract)
 			m_errorReporter.fatalTypeError(_newExpression.location(), "Identifier is not a contract.");
+		if (contract->contractKind() != ContractDefinition::ContractKind::Contract)
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Let allow it for libraries for the moment.

**Comment by axic** [comment]
@chriseth should be ready

**Comment by axic** [comment]
Life ain't easy, inheritance invokes the `newExpression` in types:

```
Internal compiler error during compilation:
/Users/alex/Projects/solidity/libsolidity/ast/Types.cpp(2147): Throw in function static FunctionTypePointer dev::solidity::FunctionType::newExpressionType(const dev::solidity::ContractDefinition &)
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: 
[dev::tag_comment*] = 
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -1543,6 +1543,8 @@ void TypeChecker::endVisit(NewExpression const& _newExpression)
 
 		if (!contract)
 			m_errorReporter.fatalTypeError(_newExpression.location(), "Identifier is not a contract.");
+		if (contract->contractKind() == ContractDefinition::ContractKind::Interface)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1561]: can you also check `isFullyImplemented`?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 3]
```diff
@@ -6655,6 +6655,32 @@ BOOST_AUTO_TEST_CASE(experimental_pragma)
 //	CHECK_ERROR_ALLOW_MULTI(text, SyntaxError, "Duplicate experimental feature name.");
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:6709]: Can this be uncommented now?

Comment (axic) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:6709]: That is being fixed in #2740.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:1561]: That is checked in the condition below this.

**Comment by axic** [comment]
@chriseth ready too

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -420,6 +420,10 @@ void TypeChecker::endVisit(InheritanceSpecifier const& _inheritance)
 	if (base->isLibrary())
 		m_errorReporter.typeError(_inheritance.location(), "Libraries cannot be inherited from.");
 
+	// Interface can have no constructors - no need to validate
+	if (base->contractKind() == ContractDefinition::ContractKind::Interface)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please add an assertion about this.

Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: Not fully sure where you want that assertion, but the `newExpressionType` below has it already and the typechecker checks for constructors not present in interfaces.

**Comment by chriseth** [comment]
Added a commit here. I think before that, the following would have been possible:

```
interface I {}
contract C is I(2) {}
```

**Comment by axic** [comment]
```
interface I {}
contract C is I(2) {}
```

This doesn't seem to be valid, there's no constructor with an uint.

**Comment by axic** [comment]
Cannot get your example to work with the old version of this PR:

```
2734.sol:3:15: Error: Wrong argument count for constructor call: 1 arguments given but expected 0.
contract C is I(2) {}
              ^--^
```

```
2734.sol:2:15: Error: Functions in interfaces cannot have an implementation.
interface I { function I(uint) {} }
              ^-----------------^
2734.sol:2:15: Error: Constructor cannot be defined in interfaces.
interface I { function I(uint) {} }
```

```
2734.sol:2:15: Error: Constructor cannot be defined in interfaces.
interface I { function I(uint); }
              ^---------------^
2734.sol:2:15: Error: Constructor must be implemented if declared.
interface I { function I(uint); }
              ^---------------^
```

**Comment by chriseth** [comment]
Hm, weird, I thought there was a "return" before control flow reaches the "Wrong argument count for constructor call" message in the old version of the PR.

**Comment by axic** [comment]
Tried again and it probably has been a stale build I've tested on. Added a test case to be sure.

**Comment by axic** [comment]
@chriseth ready to merge I think

**Review by chriseth — APPROVED** [review]

**Pull request merged**